query/sources/impl/src/main/java/org/codice/ddf/admin/sources/commons/utils/CswSourceUtils.java

```diff
@@ -0,0 +1,195 @@
+                .map(ReportWithResult::new)
+                .findFirst()
+                .orElse(createDefaultResult(hostField));
+    }
+
+    /**
+     * Attempts to create a CSW configuration from the given url.
+     *
+     * @param urlField A URL of an endpoint with CSW capabilities
+     * @param creds    optional credentials for basic authentication
+     * @return a {@link ReportWithResult} containing the {@link SourceConfigUnionField} or an {@link org.codice.ddf.admin.common.message.ErrorMessage} on failure.
+     */
+    public ReportWithResult<SourceConfigUnionField> getPreferredCswConfig(UrlField urlField, CredentialsField creds) {
+        ReportWithResult<String> responseBodyResult = requestUtils.sendGetRequest(urlField, creds,
+                GET_CAPABILITIES_PARAMS);
+
+        ReportWithResult<SourceConfigUnionField> configResult = new ReportWithResult<>();
+        if (responseBodyResult.containsErrorMsgs()) {
+            configResult.addMessages(responseBodyResult);
+            return configResult;
+        }
+
+        String requestBody = responseBodyResult.result();
+        Document capabilitiesXml;
+        try {
+            capabilitiesXml = createDocument(requestBody);
+        } catch (Exception e) {
+            LOGGER.debug("Failed to create XML document from response.");
+            configResult.argumentMessage(unknownEndpointError(urlField.path()));
+            return configResult;
+        }
+
+        CswSourceConfigurationField preferred = new CswSourceConfigurationField();
+        preferred.endpointUrl(urlField.getValue())
+                .credentials()
+                .username(creds.username())
+                .password(creds.password());
+
+        XPath xpath = XPathFactory.newInstance()
+                .newXPath();
+        xpath.setNamespaceContext(SOURCES_NAMESPACE_CONTEXT);
+
+        try {
+            if ((Boolean) xpath.compile(HAS_CATALOG_METACARD_EXP)
+                    .evaluate(capabilitiesXml, XPathConstants.BOOLEAN)) {
+                configResult.result(preferred.outputSchema(METACARD_OUTPUT_SCHEMA)
+                        .cswProfile(CSW_FEDERATION_PROFILE_SOURCE));
+                return configResult;
+            }
+        } catch (Exception e) {
+            LOGGER.debug("Failed to compile DDF Profile CSW discovery XPath expression.");
+        }
+
+        try {
+            if ((Boolean) xpath.compile(HAS_GMD_ISO_EXP)
+                    .evaluate(capabilitiesXml, XPathConstants.BOOLEAN)) {
+                configResult.result(preferred.outputSchema(GMD_OUTPUT_SCHEMA)
+                        .cswProfile(GMD_CSW_ISO_FEDERATED_SOURCE));
+                return configResult;
+            }
+        } catch (Exception e) {
+            LOGGER.debug("Failed to compile GMD CSW discovery XPath expression.");
+        }
+
+        try {
+            xpath.compile(GET_FIRST_OUTPUT_SCHEMA)
```

Comment:
Even though we don't use this result here and default to CSW 202 schema, I left it here to make sure there is an `OutputSchema` field in the response we received.